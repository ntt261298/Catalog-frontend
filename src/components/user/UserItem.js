import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';
import { PropTypes } from 'prop-types';
import { Link } from 'react-router-dom';
import {
  Button, Modal, ModalBody, ModalHeader,
} from 'reactstrap';
import { getUserItems, deleteItem } from '../../actions/item';
import { hideModal, showAddItem, showDeleteItem } from '../../actions/app';
import plus from '../../assets/images/add-button.svg';


export class UserItem extends Component {
  state = {
    categoryID: '',
    itemID: '',
  }

  componentDidMount() {
    this.props.getUserItems(this.props.token);
  }

  // componentDidUpdate(prevProps) {
  //   if(prevProps !=== this.props.item) {
  //
  //   }
  // }

  onShowDeleteItem(categoryID, itemID) {
    this.setState({
      categoryID,
      itemID,
    });
    this.props.showDeleteItem();
  }

  onDeleteItem = async () => {
    await this.props.deleteItem(this.state.categoryID, this.state.itemID, this.props.token);
    await this.props.getUserItems(this.props.token);
    this.props.hideModal();
  }

  render() {
    const {
      item, showAddItem, modal, hideModal,
    } = this.props;
    const isDeleteItemModal = modal === 'deleteItem';
    return (
      <div className="user-items">
        {
          <Fragment>
            <Modal isOpen={isDeleteItemModal} toggle={hideModal}>
              <ModalHeader toggle={hideModal}>Delete item?</ModalHeader>
              <ModalBody>
                <Button color="danger" onClick={this.onDeleteItem}>Delete</Button>
              </ModalBody>
            </Modal>
            <div className="main-user-header">
              <h2>User's Items</h2>
              <img onClick={showAddItem} src={plus} alt="" />
            </div>
            {item.userIds.map((id) => {
              const { title, description, category_id } = item.byId[id];
              return (
                <ul>
                  <li key={id} className="user-item">
                    <div className="edit-delete">
                      <span>edit</span>
                      <span
                        className="delete"
                        onClick={
                        () => this.onShowDeleteItem(category_id, id)
                      }
                      >
                        delete
                      </span>
                    </div>
                    <Link to={`/category/${category_id}/item/${id}`}>
                      <h3>{ title }</h3>
                    </Link>
                    <p>{ description }</p>
                  </li>
                </ul>
              );
            })}
          </Fragment>
        }
      </div>
    );
  }
}

UserItem.propTypes = {
  getUserItems: PropTypes.func.isRequired,
  showAddItem: PropTypes.func.isRequired,
  showDeleteItem: PropTypes.func.isRequired,
  deleteItem: PropTypes.func.isRequired,
  hideModal: PropTypes.func.isRequired,
  item: PropTypes.object.isRequired,
  token: PropTypes.string.isRequired,
  modal: PropTypes.string.isRequired,
};

const mapStateToProps = state => ({
  item: state.item,
  items: getUserItems(state),
  token: state.user.accessToken,
  modal: state.app.modal,
});

const getUserItems = state => state.item.userIds.map(id => state.item.byId[id]);

export default connect(mapStateToProps, {
  getUserItems, showAddItem, showDeleteItem, deleteItem, hideModal,
})(UserItem);
