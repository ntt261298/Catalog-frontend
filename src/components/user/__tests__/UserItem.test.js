import React from 'react';
import { shallow } from '../../../enzyme';
import { UserItem } from '../UserItem';

describe('src/components/user/UserItem', () => {
  let wrapper;
  let props;
  let deleteSpan;
  let deleteButton;

  const update = () => {
    deleteSpan = wrapper.find('.delete');
    deleteButton = wrapper.find('Button');
  };

  const setup = () => {
    wrapper = shallow(
      <UserItem {...props} />,
    );
    update();
  };

  beforeEach(() => {
    props = {
      item: {
        byId: {
          1: {
            id: 1,
            items: [],
            name: 'Item 1',
          },
          2: {
            id: 2,
            items: [],
            name: 'Item 2',
          },
        },
        allIds: [1, 2],
        userIds: [1],
        modal: '',
      },
      modal: '',
      token: 'randomToken',
      addItem: jest.fn(),
      hideModal: jest.fn(),
      getUserItems: jest.fn(),
      showAddItem: jest.fn(),
      showDeleteItem: jest.fn(),
      deleteItem: jest.fn(),
    };
  });

  it('should render add user\'s item ', () => {
    setup();
    // Expect the wrapper object to be defined
    expect(wrapper).toMatchSnapshot();
  });

  it('should call onShowDeleteItem when delete span is clicked', () => {
    setup();
    deleteSpan.simulate('click');
    expect(props.showDeleteItem).toHaveBeenCalled();
    update();
  });

  it('should call deleteItem and hideModal when delete button is clicked', () => {
    setup();
    deleteButton.simulate('click');
    expect(props.deleteItem).toHaveBeenCalled();
    expect(props.hideModal).toHaveBeenCalled();
  });
});
