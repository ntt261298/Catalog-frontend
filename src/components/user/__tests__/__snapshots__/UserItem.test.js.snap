// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/user/UserItem should render add user's item  1`] = `
<div
  className="user-items"
>
  <Modal
    autoFocus={true}
    backdrop={true}
    backdropTransition={
      Object {
        "mountOnEnter": true,
        "timeout": 150,
      }
    }
    centered={false}
    fade={true}
    isOpen={false}
    keyboard={true}
    modalTransition={
      Object {
        "timeout": 300,
      }
    }
    onClosed={[Function]}
    onOpened={[Function]}
    role="dialog"
    toggle={[MockFunction]}
    zIndex={1050}
  >
    <ModalHeader
      charCode={215}
      closeAriaLabel="Close"
      tag="h5"
      toggle={[MockFunction]}
      wrapTag="div"
    >
      Delete item?
    </ModalHeader>
    <ModalBody
      tag="div"
    >
      <Button
        color="danger"
        onClick={[Function]}
        tag="button"
      >
        Delete
      </Button>
    </ModalBody>
  </Modal>
  <div
    className="main-user-header"
  >
    <h2>
      User's Items
    </h2>
    <img
      alt=""
      onClick={[MockFunction]}
      src="test-file-stub"
    />
  </div>
  <ul>
    <li
      className="user-item"
      key="1"
    >
      <div
        className="edit-delete"
      >
        <span>
          edit
        </span>
        <span
          className="delete"
          onClick={[Function]}
        >
          delete
        </span>
      </div>
      <Link
        to="/category/undefined/item/1"
      >
        <h3 />
      </Link>
      <p />
    </li>
  </ul>
</div>
`;
